Source/WebGPU/WGSLUnitTests/WGSLLexerTests.mm

```diff
@@ -285,4 +312,155 @@ - (void) testLexerOnGraphicsShader {
+    };
+    auto checkNextTokensAreBuiltinAttr = [&](const String& attr) {
+        checkNextTokenIs(WGSL::TokenType::Attribute);
+        checkNextTokenIsIdentifier("builtin"_s);
+        checkNextTokenIs(WGSL::TokenType::ParenLeft);
+        checkNextTokenIsIdentifier(attr);
+        checkNextTokenIs(WGSL::TokenType::ParenRight);
+    };
+
+    // @vertex
+    checkNextTokenIs(WGSL::TokenType::Attribute);
+    checkNextTokenIsIdentifier("vertex"_s);
+
+    ++lineNumber;
+    // fn main(
+    checkNextTokenIs(WGSL::TokenType::KeywordFn);
+    checkNextTokenIsIdentifier("main"_s);
+    checkNextTokenIs(WGSL::TokenType::ParenLeft);
+
+    ++lineNumber;
+    //    @builtin(vertex_index) VertexIndex : u32
+    checkNextTokensAreBuiltinAttr("vertex_index"_s);
+    checkNextTokenIsIdentifier("VertexIndex"_s);
+    checkNextTokenIs(WGSL::TokenType::Colon);
+    checkNextTokenIs(WGSL::TokenType::KeywordU32);
+
+    ++lineNumber;
+    // ) -> @builtin(position) vec4<f32> {
+    checkNextTokenIs(WGSL::TokenType::ParenRight);
+    checkNextTokenIs(WGSL::TokenType::Arrow);
+    checkNextTokensAreBuiltinAttr("position"_s);
+    checkNextTokenIsIdentifier("vec4"_s);
+    checkNextTokenIs(WGSL::TokenType::LT);
+    checkNextTokenIs(WGSL::TokenType::KeywordF32);
+    checkNextTokenIs(WGSL::TokenType::GT);
+    checkNextTokenIs(WGSL::TokenType::BraceLeft);
+
+    ++lineNumber;
+    //    var pos = array<vec2<f32>, 3>(
+    checkNextTokenIs(WGSL::TokenType::KeywordVar);
+    checkNextTokenIsIdentifier("pos"_s);
+    checkNextTokenIs(WGSL::TokenType::Equal);
+    checkNextTokenIsIdentifier("array"_s);
+    checkNextTokenIs(WGSL::TokenType::LT);
+    checkNextTokenIsIdentifier("vec2"_s);
+    checkNextTokenIs(WGSL::TokenType::LT);
+    checkNextTokenIs(WGSL::TokenType::KeywordF32);
+    checkNextTokenIs(WGSL::TokenType::GT);
+    checkNextTokenIs(WGSL::TokenType::Comma);
+    checkNextTokenIsLiteral(WGSL::TokenType::IntegerLiteral, 3);
+    checkNextTokenIs(WGSL::TokenType::GT);
+    checkNextTokenIs(WGSL::TokenType::ParenLeft);
+
+    ++lineNumber;
+    //        vec2<f32>(0.0, 0.5),
+    checkNextTokenIsIdentifier("vec2"_s);
+    checkNextTokenIs(WGSL::TokenType::LT);
+    checkNextTokenIs(WGSL::TokenType::KeywordF32);
+    checkNextTokenIs(WGSL::TokenType::GT);
+    checkNextTokenIs(WGSL::TokenType::ParenLeft);
+    checkNextTokenIsLiteral(WGSL::TokenType::DecimalFloatLiteral, 0.0);
+    checkNextTokenIs(WGSL::TokenType::Comma);
+    checkNextTokenIsLiteral(WGSL::TokenType::DecimalFloatLiteral, 0.5);
+    checkNextTokenIs(WGSL::TokenType::ParenRight);
+    checkNextTokenIs(WGSL::TokenType::Comma);
+
+    ++lineNumber;
+    //        vec2<f32>(-0.5, -0.5),
+    checkNextTokenIsIdentifier("vec2"_s);
+    checkNextTokenIs(WGSL::TokenType::LT);
+    checkNextTokenIs(WGSL::TokenType::KeywordF32);
+    checkNextTokenIs(WGSL::TokenType::GT);
+    checkNextTokenIs(WGSL::TokenType::ParenLeft);
+    checkNextTokenIs(WGSL::TokenType::Minus);
```

Comment:
How is this going to work with INT_MIN?

If we lex “-3” as “minus” “3” then, for INT_MIN, it will be “minus” “some positive number that is out of bounds”

Reply:
Correct. I think there are a couple of ways to handle this:

1. The C way of specifying INT_MIN is `#define INT_MIN (-INT_MAX -1)`. Two numbers that fit in the representation of int and then an expression to generate INT_MIN.
2. The token generation doesn't convert integer literals into C ints but keeps them as strings. A rule during type checking would determine if the literal fits into the machine word.

Reply:
Oh, when I said INT_MIN I meant the literal “-2147483648“ appearing in the user’s program.

It sounds like option 1 is “force them to write “-2147483647 - 1” instead” (which sounds unfortunate :/)

Isn’t there an option 3 “combine the minus token and the integer token to make a single signed integer token at lexing time”?

Is there an option 4 “combine unary minus with an integer literal in the AST, to become a single node, before enforcing integer bounds”?

What does the spec say about this?

Reply:
I didn't mean to imply my list was exhaustive; more to illustrate that we have options. 

The spec says:

> A [shader-creation error](https://www.w3.org/TR/WGSL/#shader-creation-error) results if:
> 
> An [integer literal](https://www.w3.org/TR/WGSL/#integer-literal) with a i or u suffix cannot be represented by the target type.

Yeah, option 1 is the unfortunate, especially in a language without a pre-processor.
Any of the options is implementable and not more complicated than other other; I think it just comes down to implementation style.